File: src/verify/counterexample.ts
import type {
  TranslatorArtifact,
  ArtifactEntity,
  ArtifactStateEntry,
  ArtifactBinding,
  Z3Sort,
} from "#verify/script.js";
import { sortKey } from "#verify/script.js";
import type { Z3Model, Z3SortMap, Z3FuncMap } from "#verify/backend.js";

export interface DecodedValue {
  readonly display: string;
  readonly entityLabel: string | null;
}

export interface DecodedEntity {
  readonly sortName: string;
  readonly label: string;
  readonly rawElement: string;
  readonly fields: readonly { readonly name: string; readonly value: DecodedValue }[];
}

export interface DecodedRelationEntry {
  readonly key: DecodedValue;
  readonly value: DecodedValue;
}

export interface DecodedRelation {
  readonly stateName: string;
  readonly side: "pre" | "post";
  readonly entries: readonly DecodedRelationEntry[];
}

export interface DecodedConstant {
  readonly stateName: string;
  readonly side: "pre" | "post";
  readonly value: DecodedValue;
}

export interface DecodedInput {
  readonly name: string;
  readonly value: DecodedValue;
}

export interface CounterExample {
  readonly entities: readonly DecodedEntity[];
  readonly stateRelations: readonly DecodedRelation[];
  readonly stateConstants: readonly DecodedConstant[];
  readonly inputs: readonly DecodedInput[];
}

interface DecodeCtx {
  readonly model: Z3Model;
  readonly sortMap: Z3SortMap;
  readonly funcMap: Z3FuncMap;
  readonly artifact: TranslatorArtifact;
  readonly rawToLabel: Map<string, string>;
}

export function decodeCounterExample(
  model: Z3Model,
  sortMap: Z3SortMap,
  funcMap: Z3FuncMap,
  artifact: TranslatorArtifact,
): CounterExample {
  const ctx: DecodeCtx = { model, sortMap, funcMap, artifact, rawToLabel: new Map() };

  const entities = decodeEntities(ctx);
  const enumsDecoded = decodeEnums(ctx);
  for (const e of enumsDecoded) ctx.rawToLabel.set(e.rawElement, e.label);

  const stateRelations = decodeStateRelations(ctx);
  const stateConstants = decodeStateConstants(ctx);
  const inputs = decodeInputs(ctx);

  return { entities, stateRelations, stateConstants, inputs };
}

function decodeEntities(ctx: DecodeCtx): DecodedEntity[] {
  const result: DecodedEntity[] = [];
  for (const entity of ctx.artifact.entities) {
    const sort = ctx.sortMap.get(sortKey(entity.sort));
    if (!sort) continue;
    const universe = safeSortUniverse(ctx, sort);
    for (let i = 0; i < universe.length; i += 1) {
      const elem = universe[i];
      const raw = String(elem);
      const label = `${entity.name}#${i}`;
      ctx.rawToLabel.set(raw, label);
      const fields = decodeEntityFields(ctx, entity, elem);
      result.push({ sortName: entity.name, label, rawElement: raw, fields });
    }
  }
  return result;
}

function decodeEntityFields(
  ctx: DecodeCtx,
  entity: ArtifactEntity,
  elem: unknown,
): readonly { name: string; value: DecodedValue }[] {
  const out: { name: string; value: DecodedValue }[] = [];
  for (const field of entity.fields) {
    const decl = ctx.funcMap.get(field.funcName);
    if (!decl) continue;
    const applied = callFunc(decl, [elem]);
    const evaluated = evalExpr(ctx.model, applied);
    out.push({ name: field.name, value: decodeValueRaw(ctx, evaluated) });
  }
  return out;
}

interface DecodedEnumElement {
  readonly sortName: string;
  readonly label: string;
  readonly rawElement: string;
}

function decodeEnums(ctx: DecodeCtx): DecodedEnumElement[] {
  const result: DecodedEnumElement[] = [];
  for (const e of ctx.artifact.enums) {
    for (const member of e.members) {
      const decl = ctx.funcMap.get(member.funcName);
      if (!decl) continue;
      const applied = callFunc(decl, []);
      const evaluated = evalExpr(ctx.model, applied);
      result.push({
        sortName: e.name,
        label: `${e.name}.${member.name}`,
        rawElement: String(evaluated),
      });
    }
  }
  return result;
}

function decodeStateRelations(ctx: DecodeCtx): DecodedRelation[] {
  const result: DecodedRelation[] = [];
  for (const entry of ctx.artifact.state) {
    if (entry.kind !== "Relation") continue;
    const keySort = ctx.sortMap.get(sortKey(entry.keySort));
    const universeKeys = keySort ? safeSortUniverse(ctx, keySort) : [];
    const inputKeys = inputsOfSort(ctx, entry.keySort);
    const candidates = universeKeys.length > 0 ? universeKeys : inputKeys;
    result.push(buildRelationSide(ctx, entry, candidates, "pre"));
    if (ctx.artifact.hasPostState) {
      result.push(buildRelationSide(ctx, entry, candidates, "post"));
    }
  }
  return result;
}

function inputsOfSort(ctx: DecodeCtx, wantSort: Z3Sort): readonly unknown[] {
  const out: unknown[] = [];
  for (const b of ctx.artifact.inputs) {
    if (sortKey(b.sort) !== sortKey(wantSort)) continue;
    const decl = ctx.funcMap.get(b.funcName);
    if (!decl) continue;
    out.push(evalExpr(ctx.model, callFunc(decl, [])));
  }
  return out;
}

function buildRelationSide(
  ctx: DecodeCtx,
  relation: Extract<ArtifactStateEntry, { kind: "Relation" }>,
  keyUniverse: readonly unknown[],
  side: "pre" | "post",
): DecodedRelation {
  const domFuncName = side === "pre" ? relation.domFunc : relation.domFuncPost;
  const mapFuncName = side === "pre" ? relation.mapFunc : relation.mapFuncPost;
  const domDecl = ctx.funcMap.get(domFuncName);
  const mapDecl = ctx.funcMap.get(mapFuncName);
  const entries: DecodedRelationEntry[] = [];
  if (domDecl && mapDecl) {
    for (const k of keyUniverse) {
      const inDom = evalExpr(ctx.model, callFunc(domDecl, [k]));
      if (String(inDom) !== "true") continue;
      const mappedTo = evalExpr(ctx.model, callFunc(mapDecl, [k]));
      entries.push({
        key: decodeValueRaw(ctx, k),
        value: decodeValueRaw(ctx, mappedTo),
      });
    }
  }
  return { stateName: relation.name, side, entries };
}

function decodeStateConstants(ctx: DecodeCtx): DecodedConstant[] {
  const result: DecodedConstant[] = [];
  for (const entry of ctx.artifact.state) {
    if (entry.kind !== "Const") continue;
    result.push(buildConstantSide(ctx, entry, "pre"));
    if (ctx.artifact.hasPostState) result.push(buildConstantSide(ctx, entry, "post"));
  }
  return result;
}

function buildConstantSide(
  ctx: DecodeCtx,
  entry: Extract<ArtifactStateEntry, { kind: "Const" }>,
  side: "pre" | "post",
): DecodedConstant {
  const funcName = side === "pre" ? entry.funcName : entry.funcNamePost;
  const decl = ctx.funcMap.get(funcName);
  const value = decl
    ? decodeValueRaw(ctx, evalExpr(ctx.model, callFunc(decl, [])))
    : { display: "<unknown>", entityLabel: null };
  return { stateName: entry.name, side, value };
}

function decodeInputs(ctx: DecodeCtx): DecodedInput[] {
  return decodeBindings(ctx, ctx.artifact.inputs);
}

function decodeBindings(ctx: DecodeCtx, bindings: readonly ArtifactBinding[]): DecodedInput[] {
  const result: DecodedInput[] = [];
  for (const b of bindings) {
    const decl = ctx.funcMap.get(b.funcName);
    if (!decl) continue;
    const applied = callFunc(decl, []);
    const evaluated = evalExpr(ctx.model, applied);
    result.push({ name: b.name, value: decodeValueRaw(ctx, evaluated) });
  }
  return result;
}

function decodeValueRaw(ctx: DecodeCtx, expr: unknown): DecodedValue {
  const text = normalizeZ3Text(String(expr));
  const label = ctx.rawToLabel.get(text);
  if (label) return { display: label, entityLabel: label };
  if (text === "true" || text === "false") return { display: text, entityLabel: null };
  if (/^-?\d+$/.test(text)) return { display: text, entityLabel: null };
  const stringLit = matchStringLiteral(text);
  if (stringLit !== null) return { display: JSON.stringify(stringLit), entityLabel: null };
  return { display: prettyUninterp(text), entityLabel: null };
}

function normalizeZ3Text(raw: string): string {
  const negMatch = /^\(-\s+(\d+)\)$/.exec(raw.trim());
  if (negMatch) return `-${negMatch[1]}`;
  return raw;
}

function matchStringLiteral(text: string): string | null {
  const m = /^str_(\d+)$/.exec(text);
  return m ? `<string#${m[1]}>` : null;
}

function prettyUninterp(text: string): string {
  const m = /^([A-Za-z_][\w]*)!val!(\d+)$/.exec(text);
  return m ? `${m[1]}#${m[2]}` : text;
}

function safeSortUniverse(ctx: DecodeCtx, sort: unknown): readonly unknown[] {
  if (!isSortLike(sort)) return [];
  try {
    const universe = ctx.model.sortUniverse(sort);
    const out: unknown[] = [];
    for (let i = 0; i < universe.length(); i += 1) out.push(universe.get(i));
    return out;
  } catch {
    return [];
  }
}

function isSortLike(x: unknown): x is Parameters<Z3Model["sortUniverse"]>[0] {
  return typeof x === "object" && x !== null;
}

interface FuncLike {
  call(...args: unknown[]): unknown;
}

function isFuncLike(x: unknown): x is FuncLike {
  return typeof x === "object" && x !== null && typeof (x as FuncLike).call === "function";
}

function callFunc(decl: unknown, args: readonly unknown[]): unknown {
  if (!isFuncLike(decl)) throw new Error("func decl is not callable");
  return decl.call(...args);
}

function evalExpr(model: Z3Model, expr: unknown): unknown {
  return model.eval(expr as Parameters<Z3Model["eval"]>[0], true);
}

export function formatCounterExample(ce: CounterExample): string {
  const lines: string[] = [];
  if (ce.inputs.length > 0) {
    lines.push("  inputs:");
    for (const inp of ce.inputs) lines.push(`    ${inp.name} = ${inp.value.display}`);
  }
  if (ce.entities.length > 0) {
    lines.push("  entities:");
    for (const e of ce.entities) {
      const fieldStr = e.fields.map((f) => `${f.name} = ${f.value.display}`).join(", ");
      lines.push(`    ${e.label} { ${fieldStr} }`);
    }
  }
  const pre = ce.stateRelations.filter((r) => r.side === "pre");
  const preC = ce.stateConstants.filter((c) => c.side === "pre");
  if (pre.length > 0 || preC.length > 0) {
    lines.push("  pre-state:");
    for (const r of pre) lines.push(formatRelation(r));
    for (const c of preC) lines.push(`    ${c.stateName} = ${c.value.display}`);
  }
  const post = ce.stateRelations.filter((r) => r.side === "post");
  const postC = ce.stateConstants.filter((c) => c.side === "post");
  if (post.length > 0 || postC.length > 0) {
    lines.push("  post-state:");
    for (const r of post) lines.push(formatRelation(r));
    for (const c of postC) lines.push(`    ${c.stateName}' = ${c.value.display}`);
  }
  return lines.join("\n");
}

function formatRelation(r: DecodedRelation): string {
  const label = r.side === "pre" ? r.stateName : `${r.stateName}'`;
  if (r.entries.length === 0) return `    ${label} = {}`;
  const entryStrs = r.entries.map((e) => `${e.key.display} → ${e.value.display}`).join(", ");
  return `    ${label} = { ${entryStrs} }`;
}

File: modules/verify/src/main/scala/specrest/verify/z3/CounterExample.scala
package specrest.verify.z3

import com.microsoft.z3.Expr as Z3AstExpr
import com.microsoft.z3.FuncDecl
import com.microsoft.z3.Model
import com.microsoft.z3.RatNum
import com.microsoft.z3.Sort
import specrest.ir.generated.SpecRestGenerated.*
import specrest.verify.DecodedConstant
import specrest.verify.DecodedCounterExample
import specrest.verify.DecodedEntity
import specrest.verify.DecodedEntityField
import specrest.verify.DecodedInput
import specrest.verify.DecodedRelation
import specrest.verify.DecodedRelationEntry
import specrest.verify.DecodedValue
import specrest.verify.IrValueDecoder

import scala.collection.mutable

@SuppressWarnings(Array("org.wartremover.warts.AsInstanceOf"))
object Z3CounterExample:

  def decode(
      model: Model,
      sortMap: Map[String, Sort],
      funcMap: Map[String, FuncDecl[?]],
      artifact: TranslatorArtifact,
      ir: service_ir_full
  ): DecodedCounterExample =
    val rawToLabel  = mutable.LinkedHashMap.empty[String, String]
    val typingFails = mutable.ListBuffer.empty[String]
    val ctx         = tyctxFromService(ir)

    for e <- artifact.enums do
      for member <- e.members do
        funcMap.get(member.funcName).foreach: decl =>
          val evaluated = evalExpr(model, applyDecl(decl, Nil))
          rawToLabel(evaluated.toString) = s"${e.name}.${member.name}"

    val entities = artifact.entities.flatMap: entity =>
      val sortOpt = sortMap.get(Z3Sort.key(entity.sort))
      sortOpt match
        case None => Nil
        case Some(sort) =>
          safeSortUniverse(model, sort).zipWithIndex.map: (elem, idx) =>
            val raw   = elem.toString
            val label = s"${entity.name}#$idx"
            rawToLabel(raw) = label
            val fields = entity.fields.flatMap: field =>
              funcMap.get(field.funcName).map: decl =>
                val applied   = applyDecl(decl, List(elem))
                val evaluated = evalExpr(model, applied)
                validateType(
                  evaluated,
                  field.sort,
                  ctx,
                  rawToLabel.toMap,
                  s"entity ${entity.name}.${field.name}",
                  typingFails
                )
                DecodedEntityField(field.name, decodeValue(evaluated, rawToLabel))
            DecodedEntity(
              sortName = entity.name,
              label = label,
              rawElement = raw,
              fields = fields
            )

    val stateRelations = artifact.state.flatMap:
      case r: ArtifactStateEntry.Relation =>
        val universeKeys = sortMap.get(Z3Sort.key(r.keySort)) match
          case Some(s) => safeSortUniverse(model, s)
          case None    => Nil
        val candidates =
          if universeKeys.nonEmpty then universeKeys
          else inputsOfSort(model, funcMap, artifact.inputs, r.keySort)
        val pre =
          buildRelationSide(model, funcMap, r, candidates, "pre", rawToLabel, ctx, typingFails)
        val post = if artifact.hasPostState then
          List(buildRelationSide(
            model,
            funcMap,
            r,
            candidates,
            "post",
            rawToLabel,
            ctx,
            typingFails
          ))
        else Nil
        pre :: post
      case _: ArtifactStateEntry.Const => Nil

    val stateConstants = artifact.state.flatMap:
      case c: ArtifactStateEntry.Const =>
        val pre = buildConstantSide(model, funcMap, c, "pre", rawToLabel, ctx, typingFails)
        val post = if artifact.hasPostState then
          List(buildConstantSide(model, funcMap, c, "post", rawToLabel, ctx, typingFails))
        else Nil
        pre :: post
      case _: ArtifactStateEntry.Relation => Nil

    val inputs = artifact.inputs.flatMap: b =>
      funcMap.get(b.funcName).map: decl =>
        val evaluated = evalExpr(model, applyDecl(decl, Nil))
        validateType(
          evaluated,
          b.sort,
          ctx,
          rawToLabel.toMap,
          s"input ${b.name}",
          typingFails
        )
        DecodedInput(b.name, decodeValue(evaluated, rawToLabel))

    DecodedCounterExample(
      entities = entities,
      stateRelations = stateRelations,
      stateConstants = stateConstants,
      inputs = inputs,
      typingFailures = typingFails.toList
    )

  private def sortToTy(sort: Z3Sort, ctx: tyctx_ext[Unit]): Option[ty] = sort match
    case Z3Sort.Bool => Some(TBool())
    case Z3Sort.Int  => Some(TInt())
    case Z3Sort.Real => Some(TReal())
    case Z3Sort.Uninterp(name) =>
      if tc_enums(ctx).contains(name) then Some(TEnum(name))
      else if tc_entities(ctx)
          .exists(e => entName(e) == name)
      then
        Some(TEntity(name))
      else None
    case Z3Sort.SetOf(elem) => sortToTy(elem, ctx).map(TSet.apply)
    case Z3Sort.OptionOf(_) => None
    case Z3Sort.SeqOf(_)    => None
    case Z3Sort.Str         => None

  private def validateType(
      expr: Z3AstExpr[?],
      sort: Z3Sort,
      ctx: tyctx_ext[Unit],
      rawToLabel: Map[String, String],
      site: String,
      sink: mutable.ListBuffer[String]
  ): Unit =
    sortToTy(sort, ctx) match
      case None =>
        sink += s"$site: no ty for Z3 sort ${Z3Sort.key(sort)} (raw '${expr.toString.trim}')"
      case Some(expectedTy) =>
        IrValueDecoder.decodeZ3(expr, sort, rawToLabel) match
          case None =>
            sink += s"$site: could not decode '${expr.toString.trim}' at sort ${Z3Sort.key(sort)}"
          case Some(value) =>
            if !check_value_has_ty(ctx, value, expectedTy) then
              sink += s"$site: decoded value did not match expected type ${expectedTy}"

  private def inputsOfSort(
      model: Model,
      funcMap: Map[String, FuncDecl[?]],
      inputs: List[ArtifactBinding],
      wantSort: Z3Sort
  ): List[Z3AstExpr[?]] =
    inputs.flatMap: b =>
      if Z3Sort.key(b.sort) != Z3Sort.key(wantSort) then None
      else funcMap.get(b.funcName).map(decl => evalExpr(model, applyDecl(decl, Nil)))

  private def buildRelationSide(
      model: Model,
      funcMap: Map[String, FuncDecl[?]],
      relation: ArtifactStateEntry.Relation,
      keyUniverse: List[Z3AstExpr[?]],
      side: String,
      rawToLabel: mutable.LinkedHashMap[String, String],
      ctx: tyctx_ext[Unit],
      sink: mutable.ListBuffer[String]
  ): DecodedRelation =
    val domFuncName = if side == "pre" then relation.domFunc else relation.domFuncPost
    val mapFuncName = if side == "pre" then relation.mapFunc else relation.mapFuncPost
    val entries = (funcMap.get(domFuncName), funcMap.get(mapFuncName)) match
      case (Some(domDecl), Some(mapDecl)) =>
        keyUniverse.flatMap: k =>
          val inDom = evalExpr(model, applyDecl(domDecl, List(k)))
          if inDom.toString != "true" then None
          else
            val mappedTo = evalExpr(model, applyDecl(mapDecl, List(k)))
            validateType(
              k,
              relation.keySort,
              ctx,
              rawToLabel.toMap,
              s"relation ${relation.name}.$side key",
              sink
            )
            validateType(
              mappedTo,
              relation.valueSort,
              ctx,
              rawToLabel.toMap,
              s"relation ${relation.name}.$side value",
              sink
            )
            Some(
              DecodedRelationEntry(
                key = decodeValue(k, rawToLabel),
                value = decodeValue(mappedTo, rawToLabel)
              )
            )
      case _ => Nil
    DecodedRelation(relation.name, side, entries)

  private def buildConstantSide(
      model: Model,
      funcMap: Map[String, FuncDecl[?]],
      entry: ArtifactStateEntry.Const,
      side: String,
      rawToLabel: mutable.LinkedHashMap[String, String],
      ctx: tyctx_ext[Unit],
      sink: mutable.ListBuffer[String]
  ): DecodedConstant =
    val funcName = if side == "pre" then entry.funcName else entry.funcNamePost
    val value = funcMap.get(funcName) match
      case Some(decl) =>
        val evaluated = evalExpr(model, applyDecl(decl, Nil))
        validateType(
          evaluated,
          entry.sort,
          ctx,
          rawToLabel.toMap,
          s"constant ${entry.name}.$side",
          sink
        )
        decodeValue(evaluated, rawToLabel)
      case None => DecodedValue("<unknown>", None)
    DecodedConstant(entry.name, side, value)

  // Z3 returns Real model values as RatNum; render them readably (n or n/d).
  private def formatReal(expr: Z3AstExpr[?]): Option[String] = expr match
    case r: RatNum =>
      val num = r.getBigIntNumerator
      val den = r.getBigIntDenominator
      Some(if den.compareTo(java.math.BigInteger.ONE) == 0 then num.toString else s"$num/$den")
    case _ => None

  private def decodeValue(
      expr: Z3AstExpr[?],
      rawToLabel: mutable.LinkedHashMap[String, String]
  ): DecodedValue =
    formatReal(expr) match
      case Some(r) => DecodedValue(r, None)
      case None =>
        val text = normalizeZ3Text(expr.toString)
        rawToLabel.get(text) match
          case Some(label) => DecodedValue(label, Some(label))
          case None =>
            if text == "true" || text == "false" then DecodedValue(text, None)
            else if text.matches("-?\\d+") then DecodedValue(text, None)
            else
              matchStringLiteral(text) match
                case Some(s) => DecodedValue(s"\"$s\"", None)
                case None    => DecodedValue(prettyUninterp(text), None)

  private val NegNum       = """^\(-\s+(\d+)\)$""".r
  private val StrLit       = """^str_(\d+)$""".r
  private val UninterpName = """^([A-Za-z_][\w]*)!val!(\d+)$""".r

  private def normalizeZ3Text(raw: String): String =
    NegNum.findFirstMatchIn(raw.trim) match
      case Some(m) => s"-${m.group(1)}"
      case None    => raw

  private def matchStringLiteral(text: String): Option[String] =
    StrLit.findFirstMatchIn(text).map(m => s"<string#${m.group(1)}>")

  private def prettyUninterp(text: String): String =
    UninterpName.findFirstMatchIn(text) match
      case Some(m) => s"${m.group(1)}#${m.group(2)}"
      case None    => text

  private def safeSortUniverse(model: Model, sort: Sort): List[Z3AstExpr[?]] =
    scala.util.Try(model.getSortUniverse(sort).toList).toOption.getOrElse(Nil)

  private def applyDecl(decl: FuncDecl[?], args: List[Z3AstExpr[?]]): Z3AstExpr[?] =
    decl
      .asInstanceOf[FuncDecl[Sort]]
      .apply(args.toArray.asInstanceOf[Array[Z3AstExpr[Sort]]]*)

  private def evalExpr(model: Model, expr: Z3AstExpr[?]): Z3AstExpr[?] =
    model.eval(expr.asInstanceOf[Z3AstExpr[Sort]], true)
